backends/tfhe-cuda-backend/cuda/include/integer.h

```diff
@@ -414,93 +431,110 @@ template <typename Torus> struct int_radix_lut {
   uint32_t num_blocks;
   bool mem_reuse = false;
 
-  int8_t *buffer;
+  // There will be one buffer on each GPU in multi-GPU computations
+  // (same for tmp lwe arrays)
+  std::vector<int8_t *> buffer;
 
+  // These arrays will all reside on GPU 0
+  // lut could actually be allocated & initialized GPU per GPU but this is not
+  // done at the moment
+  Torus *lut = nullptr;
   Torus *lut_indexes;
+  // All tmp lwe arrays and index arrays for lwe contain the total
+  // amount of blocks to be computed on, there is no split between GPUs
+  // for the moment
   Torus *lwe_indexes_in;
   Torus *lwe_indexes_out;
-
   // lwe_trivial_indexes is the intermediary index we need in case
   // lwe_indexes_in != lwe_indexes_out
   Torus *lwe_trivial_indexes;
-
   Torus *tmp_lwe_before_ks;
   Torus *tmp_lwe_after_ks;
 
-  Torus *lut = nullptr;
-
-  int_radix_lut(cudaStream_t stream, uint32_t gpu_index,
-                int_radix_params params, uint32_t num_luts,
+  int_radix_lut(cudaStream_t *streams, uint32_t *gpu_indexes,
+                uint32_t gpu_count, int_radix_params params, uint32_t num_luts,
                 uint32_t num_radix_blocks, bool allocate_gpu_memory) {
 
     this->params = params;
     this->num_blocks = num_radix_blocks;
     Torus lut_indexes_size = num_radix_blocks * sizeof(Torus);
-    Torus big_size =
-        (params.big_lwe_dimension + 1) * num_radix_blocks * sizeof(Torus);
-    Torus small_size =
-        (params.small_lwe_dimension + 1) * num_radix_blocks * sizeof(Torus);
     Torus lut_buffer_size =
         (params.glwe_dimension + 1) * params.polynomial_size * sizeof(Torus);
 
     ///////////////
-    execute_scratch_pbs<Torus>(
-        stream, gpu_index, &buffer, params.glwe_dimension,
-        params.small_lwe_dimension, params.polynomial_size, params.pbs_level,
-        params.grouping_factor, num_radix_blocks,
-        cuda_get_max_shared_memory(gpu_index), params.pbs_type,
-        allocate_gpu_memory);
+    auto active_gpu_count = get_active_gpu_count(num_radix_blocks, gpu_count);
+    cuda_synchronize_stream(streams[0], gpu_indexes[0]);
+    for (uint i = 0; i < active_gpu_count; i++) {
+      cudaSetDevice(i);
+      int8_t *gpu_pbs_buffer;
+      auto num_blocks_on_gpu =
+          get_num_inputs_on_gpu(num_radix_blocks, i, gpu_count);
+
+      execute_scratch_pbs<Torus>(
+          streams[i], gpu_indexes[i], &gpu_pbs_buffer, params.glwe_dimension,
+          params.small_lwe_dimension, params.polynomial_size, params.pbs_level,
+          params.grouping_factor, num_blocks_on_gpu,
+          cuda_get_max_shared_memory(gpu_indexes[i]), params.pbs_type,
+          allocate_gpu_memory);
+      cuda_synchronize_stream(streams[i], gpu_indexes[i]);
+      buffer.push_back(gpu_pbs_buffer);
+    }
```

Comment:
don't we need cpu multi-threading here?

Reply:
No if we do we'll push_back in the wrong order.